modules: mbedtls: align Kconfig names to Mbed TLS build symbols

valeriosetti opened the pull request:
This is a first step toward having a 1:1 mapping between Kconfig used in Zephyr and the build symbols used in Mbed TLS to ease maintenance.

Even though it might seem that a lot of files are changed it really is just a big renaming PR.
I started by creating a script that reads all the Kconfig names from `zephyr/modules/mbedtls/Kconfig.mbedtls` and then verifies if they are present in Mbed TLS official configuration file `modules/crypto/mbedtls/include/mbedtls/mbedtls_config.h`. All the unmatched Kconfig were split into 3 categories:
- "keep as is": symbols that are going to be removed in the next Mbed TLS release (4.0) are not worth to be renamed now since that release is expected soon (likely October 2025);
- "rename": rename using the Mbed TLS format;
- "not Mbed TLS symbols": Kconfigs that are Zephyr customizations and that don't have an Mbed TLS counterpart.

Here's the list that I got:
```
Don't touch - they will be removed in 4.0/1.0
---------------------------------------------
MBEDTLS_CMAC -> MBEDTLS_CMAC_C
MBEDTLS_MD5 -> MBEDTLS_MD5_C
MBEDTLS_SHA1 -> MBEDTLS_SHA1_C
MBEDTLS_SHA224 -> MBEDTLS_SHA224_C
MBEDTLS_SHA256 -> MBEDTLS_SHA256_C
MBEDTLS_SHA384 -> MBEDTLS_SHA384_C
MBEDTLS_SHA512 -> MBEDTLS_SHA512_C
MBEDTLS_POLY1305 -> MBEDTLS_POLY1305_C
MBEDTLS_CIPHER -> MBEDTLS_CIPHER_C
MBEDTLS_CIPHER_AES_ENABLED -> MBEDTLS_AES_C
MBEDTLS_CIPHER_MODE_XTS_ENABLED -> MBEDTLS_CIPHER_MODE_XTS
MBEDTLS_CIPHER_CAMELLIA_ENABLED -> MBEDTLS_CAMELLIA_C
MBEDTLS_CIPHER_DES_ENABLED -> MBEDTLS_DES_C
MBEDTLS_CIPHER_CHACHA20_ENABLED -> MBEDTLS_CHACHA20_C
MBEDTLS_CHACHAPOLY_AEAD_ENABLED -> MBEDTLS_CHACHAPOLY_C
MBEDTLS_CIPHER_CCM_ENABLED -> MBEDTLS_CCM_C
MBEDTLS_CIPHER_GCM_ENABLED -> MBEDTLS_GCM_C
MBEDTLS_CIPHER_MODE_CBC_ENABLED -> MBEDTLS_CIPHER_MODE_CBC
MBEDTLS_CIPHER_MODE_CTR_ENABLED -> MBEDTLS_CIPHER_MODE_CTR
MBEDTLS_GENPRIME_ENABLED -> MBEDTLS_GENPRIME

Rename
------
MBEDTLS_MD -> MBEDTLS_MD_C
MBEDTLS_LMS -> MBEDTLS_LMS_C
MBEDTLS_TLS_VERSION_1_2 -> MBEDTLS_SSL_PROTO_TLS1_2
MBEDTLS_DTLS -> MBEDTLS_SSL_PROTO_DTLS
MBEDTLS_TLS_VERSION_1_3 -> MBEDTLS_SSL_PROTO_TLS1_3
MBEDTLS_TLS_SESSION_TICKETS -> MBEDTLS_SSL_SESSION_TICKETS
MBEDTLS_CTR_DRBG_ENABLED -> MBEDTLS_CTR_DRBG_C
MBEDTLS_HMAC_DRBG_ENABLED -> MBEDTLS_HMAC_DRBG_C

Not Mbed TLS symbols
--------------------
MBEDTLS_KEY_EXCHANGE_ALL_ENABLED
MBEDTLS_ECP_ALL_ENABLED
MBEDTLS_CIPHER_ALL_ENABLED
MBEDTLS_HASH_ALL_ENABLED
MBEDTLS_PEM_CERTIFICATE_FORMAT
MBEDTLS_ENTROPY_POLL_ZEPHYR
MBEDTLS_OPENTHREAD_OPTIMIZATIONS_ENABLED
MBEDTLS_USER_CONFIG_ENABLE
MBEDTLS_PSA_CRYPTO_LEGACY_RNG 
MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG_ALLOW_NON_CSPRNG 
MBEDTLS_PSA_P256M_DRIVER_RAW 
```

In conclusion Kconfig which are renamed in this PR are those belonging to the "rename" section above.

For symbols belonging to the "don't touch" area I'm planning to open PR(s) to remove them from the codebase replacing them with PSA API.

dkalowsk reviewed (dismissed):
I like this.  Makes sync's with mbedTLS a lot easier.  Thanks!

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_MD

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_TLS_VERSION_1_2

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_DTLS

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_TLS_VERSION_1_3

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_TLS_SESSION_TICKETS

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_CTR_DRBG_ENABLED

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_HMAC_DRBG_ENABLED

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
modules: mbedtls: rename MBEDTLS_LMS

Align Kconfig name to the Mbed TLS build symbol.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

rlubos reviewed (commented):
I think this requires an entry in the migration guide.

valeriosetti committed:
modules: mbedtls: add a specific file for deprecated Kconfigs

This is to preserve backward compatibility until the deprecation period of
the old names expires. After that time entries from that file can be
removed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
doc: migration-guide: add note for Mbed TLS' Kconfigs renaming

Some Mbed TLS' Kconfigs were renamed in order to improve the 1:1 matching
between them and the build symbols used in Mbed TLS. This commit add a note
on the migration-guide document for this change.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.deprecated:
```diff
@@ -0,0 +1,60 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config MBEDTLS_MD
+	bool
```
https://github.com/zephyrproject-rtos/zephyr/pull/92081/commits/c0ebe4adcd1d1bef1c104c4bea07b80cd9d05375#r2343664592

tomi-font commented on modules/mbedtls/Kconfig.deprecated:
```diff
@@ -0,0 +1,60 @@
+# user selectable, but they are kept for backward compatibility until the
+# deprecation period expires.
+#
+# Copyright (c) 2025 BayLibre SAS
+# SPDX-License-Identifier: Apache-2.0
```
Put copyright & license as first thing in the file?

tomi-font commented on modules/mbedtls/Kconfig.deprecated:
```diff
@@ -0,0 +1,45 @@
+# Deprecated Mbed TLS symbols. No prompt is added because they shouldn't be
+# user selectable, but they are kept for backward compatibility until the
+# deprecation period expires.
```
The whole comment could now be removed.

tomi-font commented on modules/mbedtls/configs/config-mbedtls.h:
```diff
@@ -41,8 +41,8 @@
-#if defined(CONFIG_MBEDTLS_LMS)
-#define MBEDTLS_LMS_C
+#if defined($MBEDTLS_LMS)
+#define $MBEDTLS_LMS
```
fail

nordicjm replied:
I'm not really following what this even does, if the variable is defined, it defines the same variable... And it's got a `$` as the first character which looks like a mistake? And it doesn't have a `CONFIG_` prefix in the `if` which looks like a mistake?

tomi-font commented on modules/mbedtls/configs/config-mbedtls.h:
```diff
@@ -41,8 +41,8 @@
 
-#if defined($MBEDTLS_LMS)
-#define $MBEDTLS_LMS
+#if defined(MBEDTLS_LMS_C)
```
`CONFIG_MBEDTLS_LMS_C`

tomi-font commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -1279,11 +1279,11 @@ Libraries / Subsystems
     secure random sources when :kconfig:option:`CONFIG_MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`
     is also enabled. This is only meant to be used for test purposes, not in production.
     (:github:`76408`)
-  * The Kconfig symbol :kconfig:option:`CONFIG_MBEDTLS_TLS_VERSION_1_3` was added to
```
At least right now that you added back the deprecated Kconfig options you shouldn't _have_ to make any changes here. But when they are eventually removed probably the problem will come up again.

To me it seems more right to remove the link than to rename the Kconfig option. Even though the release note remains correct and applicable for Zephyr 4.3, it is not for Zephyr 4.0. For someone using a Zephyr release before 4.3 `CONFIG_MBEDTLS_SSL_PROTO_TLS1_3` won't actually exist (yet).

valeriosetti replied:
Sorry, I completely missed the fact that I changed an old release note file. That's definitely not right and I will revert this.

tomi-font commented on modules/mbedtls/configs/config-mbedtls.h:
```diff
@@ -65,14 +65,14 @@
 
 #if defined(CONFIG_MBEDTLS_SSL_PROTO_TLS1_2) || \
-	defined(CONFIG_MBEDTLS_TLS_VERSION_1_3)
+	defined(CONFIG_MBEDTLS_SSL_PROTO_TLS1_3)
```
This fits on a single line, combine them?

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -182,6 +182,20 @@ Logging
+Mbed TLS
+========
+
+* In order to improve the 1:1 matching between Zephyr Kconfig and Mbed TLS build symbols, following
```
```suggestion
* In order to improve the 1:1 matching between Zephyr Kconfig and Mbed TLS build symbols, the following
```

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -182,6 +182,20 @@ Logging
+========
+
+* In order to improve the 1:1 matching between Zephyr Kconfig and Mbed TLS build symbols, following
+  Kconfigs were renamed:
+  * :kconfig:option:`MBEDTLS_MD` -> :kconfig:option:`MBEDTLS_MD_C`
```
This looks wrong: https://builds.zephyrproject.io/zephyr/pr/92081/docs/releases/migration-guide-4.3.html#mbed-tls
I believe you need an empty line here:
```suggestion
  Kconfigs were renamed:

  * :kconfig:option:`MBEDTLS_MD` -> :kconfig:option:`MBEDTLS_MD_C`
```

tomi-font replied:
^

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -182,6 +182,20 @@ Logging
+
+* In order to improve the 1:1 matching between Zephyr Kconfig and Mbed TLS build symbols, following
+  Kconfigs were renamed:
+  * :kconfig:option:`MBEDTLS_MD` -> :kconfig:option:`MBEDTLS_MD_C`
+  * :kconfig:option:`MBEDTLS_LMS` -> :kconfig:option:`MBEDTLS_LMS_C`
+  * :kconfig:option:`MBEDTLS_TLS_VERSION_1_2` -> :kconfig:option:`MBEDTLS_SSL_PROTO_TLS1_2`
+  * :kconfig:option:`MBEDTLS_DTLS` -> :kconfig:option:`MBEDTLS_SSL_PROTO_DTLS`
+  * :kconfig:option:`MBEDTLS_TLS_VERSION_1_3` -> :kconfig:option:`MBEDTLS_SSL_PROTO_TLS1_3`
+  * :kconfig:option:`MBEDTLS_TLS_SESSION_TICKETS` -> :kconfig:option:`MBEDTLS_SSL_SESSION_TICKETS`
+  * :kconfig:option:`MBEDTLS_CTR_DRBG_ENABLED` -> :kconfig:option:`MBEDTLS_CTR_DRBG_C`
+  * :kconfig:option:`MBEDTLS_HMAC_DRBG_ENABLED` -> :kconfig:option:`MBEDTLS_HMAC_DRBG_C`
```
All those should be prefixed with `CONFIG_`.

tomi-font reviewed (changes requested):
Nice work and well structured! Just some minor things to fix.

Regarding the `"keep as is": symbols that are going to be removed in the next Mbed TLS release (4.0) are not worth to be renamed now since that release is expected soon (likely October 2025)`: I'm wondering if we should mark them as deprecated to warn users?

tomi-font commented on modules/mbedtls/Kconfig.deprecated:
```diff
@@ -0,0 +1,58 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config MBEDTLS_MD
+	bool "[DEPRECATED] Generic message digest layer."
```
```suggestion
	bool "[DEPRECATED] Generic message digest layer"
```
? 🙂 if you need to make other changes

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -227,6 +216,21 @@ Logging
+========
+
+* In order to improve the 1:1 matching between Zephyr Kconfig and Mbed TLS build symbols, the
+  following Kconfigs were renamed:
```
Actually they weren't renamed, just new ones were introduced... But we can keep this simple, as it is, as anyway we want the users to switch to the new symbols sooner or later.

valeriosetti committed:
modules: mbedtls: fix prompts and help messages for Kconfigs

Minor fixes to prompt and help messages for some of the Mbed TLS' Kconfigs.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

nordicjm commented on modules/mbedtls/Kconfig.mbedtls:
```diff
@@ -374,22 +374,22 @@ endmenu
 
-config MBEDTLS_MD
+config MBEDTLS_MD_C
 	bool "generic message digest layer."
```
This isn't fixed in the correct commit? Fix in the correct commit
